File: tests/colour_test.js
// Pure tests for the tag-colour helpers. Run with: node tests/colour_test.js
// colour.js has no DOM/host dependencies, so plain node is enough — same
// pattern as tab_index_test.js.
const {
  COLOUR_DEFAULT, colourVar, tagColourSlot, colourSignature,
} = require("../ui/colour.js");

let failures = 0;
function check(label, got, want) {
  if (got === want) {
    console.log("ok   " + label);
  } else {
    failures++;
    console.log("FAIL " + label + " — got " + String(got) + ", want " + String(want));
  }
}

// --- colourVar -------------------------------------------------------
// Slot 0 is a real colour (Black) and must not be treated as "unset".
check("colourVar null -> link",    colourVar(null),           "var(--link)");
check("colourVar undefined",       colourVar(undefined),      "var(--link)");
check("colourVar DEFAULT",         colourVar(COLOUR_DEFAULT), "var(--link)");
check("colourVar 0 is a colour",   colourVar(0),  "var(--ansi-0, var(--link))");
check("colourVar 15",              colourVar(15), "var(--ansi-15, var(--link))");

// --- tagColourSlot ---------------------------------------------------
const settings = {
  sources: { tells: { colour: 4 }, group: { colour: 0 } },
  channels: {
    Wizards:  { colour: 9 },
    Newbie:   {},              // in the registry, no colour chosen
    Zero:     { colour: 0 },
  },
};

check("tells comes off the source",  tagColourSlot(settings, "tells", null), 4);
check("group comes off the source",  tagColourSlot(settings, "group", "AdHocGroupName"), 0);
check("channel line -> registry",    tagColourSlot(settings, "channels", "Wizards"), 9);
check("pinned tab -> registry",      tagColourSlot(settings, "channel:Wizards", "Wizards"), 9);
check("channel with no colour",      tagColourSlot(settings, "channels", "Newbie"), null);
check("channel slot 0 kept",         tagColourSlot(settings, "channels", "Zero"), 0);
check("unknown channel",             tagColourSlot(settings, "channels", "Nope"), null);
check("no channel, no source",       tagColourSlot(settings, "all", null), null);

// A group line's tag carries the group's ad hoc name; it must resolve to the
// group source even when a channel of that name happens to be registered.
check("group beats a same-named channel",
  tagColourSlot(settings, "group", "Wizards"), 0);

// Missing/partial settings must not throw — the panel renders replayed lines
// before the first `settings` broadcast arrives.
check("empty settings",  tagColourSlot({}, "channels", "Wizards"), null);
check("bare tells",      tagColourSlot({}, "tells", null), null);

// --- colourSignature -------------------------------------------------
const sig = colourSignature;
const base = { sources: { tells: {}, group: {} }, channels: {} };

check("signature is stable",
  sig(base) === sig({ sources: { tells: {}, group: {} }, channels: {} }), true);

check("channel colour changes it",
  sig(base) !== sig({ ...base, channels: { A: { colour: 1 } } }), true);

check("source colour changes it",
  sig(base) !== sig({ sources: { tells: { colour: 1 }, group: {} }, channels: {} }), true);

check("clearing a colour changes it",
  sig({ ...base, channels: { A: { colour: 1 } } }) !== sig({ ...base, channels: { A: {} } }), true);

check("slot 0 is not 'unset'",
  sig({ ...base, channels: { A: { colour: 0 } } }) !== sig(base), true);

// Non-colour churn must NOT change it — this is the whole point of the
// signature, since every settings broadcast would otherwise repaint (and
// re-anchor) the scrollback.
check("last_seen/count churn is ignored",
  sig({ ...base, channels: { A: { colour: 1, count: 1, last_seen: 100 } } })
  === sig({ ...base, channels: { A: { colour: 1, count: 99, last_seen: 999 } } }), true);

check("toggles are ignored",
  sig({ ...base, channels: { A: { colour: 1, pinned: false } } })
  === sig({ ...base, channels: { A: { colour: 1, pinned: true } } }), true);

check("a new uncoloured channel is ignored",
  sig({ ...base, channels: { A: { colour: 1 } } })
  === sig({ ...base, channels: { A: { colour: 1 }, B: {} } }), true);

// Key order must not matter: Object.entries follows insertion order, and the
// registry is rebuilt from JSON on every broadcast.
check("key order does not matter",
  sig({ ...base, channels: { A: { colour: 1 }, B: { colour: 2 } } })
  === sig({ ...base, channels: { B: { colour: 2 }, A: { colour: 1 } } }), true);

// Sources and channels are namespaced, so a channel named "tells" can't
// forge the tells source entry and mask a real change.
check("channel named 'tells' does not collide with the source",
  sig({ sources: { tells: { colour: 3 }, group: {} }, channels: {} })
  !== sig({ sources: { tells: {}, group: {} }, channels: { tells: { colour: 3 } } }), true);

console.log("");
if (failures === 0) {
  console.log("all passed");
} else {
  console.log(failures + " failure(s)");
  process.exit(1);
}

File: tests/colour_test.lua
-- Pure-Lua tests for colour.sanitise / colour.resolve.
--
-- Run with: lua tests/colour_test.lua
--
-- colour.lua has no host-API dependencies, so a vanilla Lua interpreter
-- is enough — same pattern as classifier_test.lua / flush_gate_test.lua.
--
-- This is the settings_update trust boundary: everything here arrives from
-- the panel iframe as decoded JSON, so the point of these cases is that
-- nothing but an in-range integer ever reaches storage.

package.path = "src/?.lua;" .. package.path
local colour = require("colour")

local failures = 0
local function check(label, got, want)
  if got == want then
    print("ok   " .. label)
  else
    failures = failures + 1
    print("FAIL " .. label .. " — got " .. tostring(got) .. ", want " .. tostring(want))
  end
end

-- sanitise: the valid range, inclusive at both ends.
check("sanitise 0 (Black)",   colour.sanitise(0),  0)
check("sanitise 7",           colour.sanitise(7),  7)
check("sanitise 15 (last)",   colour.sanitise(15), 15)

-- sanitise: out of range.
check("sanitise -1",          colour.sanitise(-1),  nil)
check("sanitise 16",          colour.sanitise(16),  nil)
check("sanitise -100",        colour.sanitise(-100), nil)

-- sanitise: floats truncate toward the floor rather than being rejected,
-- so a JSON number that decoded as 3.0 still works.
check("sanitise 3.0",         colour.sanitise(3.0), 3)
check("sanitise 3.7 floors",  colour.sanitise(3.7), 3)
check("sanitise -0.5 floors out of range", colour.sanitise(-0.5), nil)

-- sanitise: non-numbers.
check("sanitise nil",         colour.sanitise(nil),      nil)
check("sanitise string",      colour.sanitise("3"),      nil)
check("sanitise bool",        colour.sanitise(true),     nil)
check("sanitise table",       colour.sanitise({}),       nil)

-- sanitise: NaN. Every comparison against NaN is false, so a plain range
-- check passes it through; it must be rejected explicitly or we store a
-- value that can't be JSON-encoded on the way back to the panel.
local nan = 0 / 0
check("sanitise NaN",         colour.sanitise(nan),  nil)
check("sanitise +inf",        colour.sanitise(math.huge),  nil)
check("sanitise -inf",        colour.sanitise(-math.huge), nil)

-- resolve: the DEFAULT sentinel clears, and so does anything invalid —
-- garbage resets the row rather than wedging it.
check("resolve DEFAULT clears",  colour.resolve(colour.DEFAULT), nil)
check("resolve -1 is DEFAULT",   colour.resolve(-1),   nil)
check("resolve valid slot",      colour.resolve(9),    9)
check("resolve slot 0",          colour.resolve(0),    0)
check("resolve out of range",    colour.resolve(99),   nil)
check("resolve string",          colour.resolve("red"), nil)
check("resolve NaN",             colour.resolve(nan),  nil)

-- Guards the `and`/`or` trap the original inline version relied on: slot 0
-- is falsy in most languages but truthy in Lua, so it must survive resolve
-- as 0 and not collapse to nil.
check("slot 0 survives resolve (not nil)", colour.resolve(0) ~= nil, true)

if failures == 0 then
  print("all colour tests passed")
else
  print(failures .. " failure(s)")
  os.exit(1)
end
